Register client as browser's default magnet handler

mentalblank opened the pull request:
This PR introduces a new button to the navbar labelled "🧲 Register Magnet Handler" which uses `navigator.registerProtocolHandler` to register the client as a handler for `magnet:` links.

Magnet links will take users to the client's "Add New Torrent" page, automatically populating the "Magnet Link" input box, and allowing users to modify all options before adding the torrent to the queue. (I did consider bypassing this at one point and directly adding it to the queue with the GUI default settings, but simplicity and flexibility seem preferable.)

Since protocol handler registration requires the page to be served over a secure connection, I’ve added a check to ensure the site is loaded over HTTPS (`window.location.protocol !== "https:"`). If not, the user is shown an alert explaining that a secure context is required.

[MDN's browser compatibility chart](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler#browser_compatibility), states this feature is supported in:
- Chrome 30+
- Edge 79+
- Firefox 2+
- Opera 17+
- Firefox for Android 5+

This primarily makes it only useful for desktop users or anyone with an Android device. I have tested it on Chrome, Edge, Firefox, and Android.

I should note that the button was placed in the navigation bar for personal convenience, as I swap between several applications, but I’m open to moving it elsewhere. Perhaps it would be useful as part of the initial setup flow?

rogerfar commented:
Very cool, didn't know this was possible. I think the Settings page would be a better place to put the setting? Maybe just on the bottom of the General page?

mentalblank commented:
I have just pushed a commit that moves the handler registration button to the bottom of the general tab. To keep in line with the tab's existing formatting, I've added a small help note.

I've made the class `button is-info` to prevent visual confusion between this and the Save Settings button. Other classes like `is-warning` didn't seem like the right fit.

![Screenshot 2025-05-02 182242](https://github.com/user-attachments/assets/dc14a684-6764-440c-bbd8-8ebdb8d2285e)

Cucumberrbob commented on client/src/app/settings/settings.component.ts:
```diff
@@ -139,4 +139,20 @@ export class SettingsComponent implements OnInit {
   }
+  
+  public registerMagnetHandler(): void {
+    if (window.location.protocol !== "https:") {
```
[Secure Context](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts) doesn’t necessarily mean https
localhost is also considered a secure context, so it would be nice to let users who run rdt-client on the machine they browse on use this feature too.

we can use [`window.isSecureContext`](https://developer.mozilla.org/en-US/docs/Web/API/Window/isSecureContext)
```suggestion
    if (!window.isSecureContext) {
```

mentalblank commented:
I've made a few additional tweaks in addition to the suggested changes.

- Removed HTTPS protocol check in `registerMagnetHandler()`. Now using `this.canRegisterMagnetHandler = !!(window.isSecureContext && 'registerProtocolHandler' in navigator);` to determine support for protocol handler registration in a secure context.
- Updated the registration button and help text to respond dynamically based on `canRegisterMagnetHandler` and provide a better user experience and messaging when unsupported.
- Wrapped `navigator.registerProtocolHandler()` in a try-catch block to handle registration failures gracefully and avoid unhandled exceptions.

This allows the button to be disabled automatically, or catching of errors, on outdated/forked browsers that do not support localhost as a secure context, or where protocol handler registration is unavailable, or a few edge cases that would prevent it from registering (eg. when FF's `network.protocol-handler.external.magnet = false` or other user changes).

Here's an example of this from a user's perspective. (left = Firefox fork, right = Chrome on Android)

![FFvChrome](https://github.com/user-attachments/assets/c72c91bf-62c8-476e-b494-46b17b95c4ff)

Cucumberrbob reviewed (approved):
Fantastic work! Thanks again for your contribution!

I've tested this with Firefox over https and it works.
